Adding JunitXML output + minor reorg

csibbitt opened the pull request:
* Adds a custom wrapper to the built-in junit callback to help with naming and config
* Extracts tests into their own taskfiles

csibbitt committed:
Adding JunitXML output

csibbitt committed:
Orginize tests into their own taskfiles

csibbitt commented:
## Testing
Here is the XML that is produces

```
<?xml version="1.0" ?>
<testsuites disabled="0" errors="0" failures="0" tests="2" time="2.1332314014434814453125">
	<testsuite disabled="0" errors="0" failures="0" name="main" skipped="0" tests="2" time="2.1332314014434814453125">
		<testcase classname="rhoso_rhelai_validation.gpus_tests" name="check_if_gpus_in_passthrough_mode_are_present_in_rhel_ai_vm_lspci" time="2.1073696613311768"/>
		<testcase classname="rhoso_rhelai_validation.nvidia_tests" name="check_if_gpus_in_passthrough_mode_are_present_in_rhel_ai_vm_nvidia_smi" time="0.025861740112304688"/>
	</testsuite>
</testsuites>
```

csibbitt commented:
@lpiwowar what do you think of this organization?

The majority of existing tasks can be seen as "prep" tasks that will mutate the system and gather information. If these fail, testing stops.

The "test" tasks check on the results of the preparation, but do not mutate the system or define any variables for other tasks. Since they are orthogonal to any other tasks, we can ignore errors in order to maximize the testing feedback.

This should make it easy to track down tests rather than looking for them in a long list of tasks, which is something you mentioned being concerned about. The downside is that you basically will need to read the prep tasks anyways in order to understand them.

csibbitt commented on callback_plugins/custom_junit.py:
```diff
@@ -0,0 +1,34 @@
+        tc.name = self.mutate_task_name(tc.name)
+        tc.system_out = None
+        tc.system_err = None
+        tc.classname = 'rhoso_rhelai_validation.' + re.sub(r'\.yaml:[0-9]+$', '', os.path.basename(task_data.path))
```
I don't really like this classname system (that I just added). It means that the classname will change if we reorganize our ansible code, which doesn't seem ideal to me (though I know historically that's what classname means). Instead I was thinking of something like "TEST[gpus]"  "TEST[nvidia]", etc and I could parse that class name out to add to `rhoso_rhelai_validation.`

Thoughts?

lpiwowar replied:
Just for context. For example, in Tempest the test name changes if we move the test code to a different folder / file, but we have a UUID assigned to each test that allows you to track the test and see how it changed throughout time [1]. But unfortunately, the UUID is not used in Polarion AFAIK.

I'm just thinking aloud here. Another option might be adding the full name into the test task itself:
```
TEST rhoso_rhelai_validation.gpus_tests.check_if_gpus_in_passthrough_mode_are_present_in_rhel_ai_vm_lspc
```

This way the name would be more coupled with the test itself, and it is less likely that the name of the test would change when it is moved around. 

Maybe that was what you meant by your comment. I'm not sure.

[1] https://opendev.org/openstack/tempest/src/commit/faa1decedfef73a77ab88feee2976036da475e0d/tempest/api/volume/test_image_metadata.py#L42

csibbitt replied:
Output snippet for context:
```
<testcase classname="rhoso_rhelai_validation.nvidia_tests" name="check_if_gpus_in_passthrough_mode_are_present_in_rhel_ai_vm_nvidia_smi" time="0.025861740112304688"/>
```

I would like it if the classname + name uniquely identified the test, wherever it may move in a refactored codebase (like your UUID). For example, if we decide naming our taskfiles `*_asserts.yaml` is better than `*_tests.yaml` I don't want to have to adjust any downstream testing systems. Right now the classname suffix depends on the path, so while I don't think we need to "add the full name into the test task", explicitly adding the classname suffix seems like a good idea. I'll stop arguing against my own diff and instead code up my proposed improvement. :smile:

csibbitt commented:
Adding @jpodivin because I see he's assigned some work to do the next step of Polarion integration.

lpiwowar commented on rhelai-validation/tasks/main.yaml:
```diff
@@ -2,10 +2,16 @@
+- name: Test GPUs  # noqa: ignore-errors
+  ansible.builtin.import_tasks: gpus_tests.yaml
+  ignore_errors: true
+- name: Prep NVIDIA
+  ansible.builtin.import_tasks: nvidia_prep.yaml
+- name: Test NVIDIA  # noqa: ignore-errors
+  ansible.builtin.import_tasks: nvidia_tests.yaml
+  ignore_errors: true
```
**question:** Would it make sense to rename the prep section to something else?

I like how it is structured. This is a subjective point of view, but I do not like the name prep. When I read this without investigating the playbooks, `prep` implies to me that it prepares the environment for the testing when in reality it is already performing the testing from my point of view. I would personally say `nvidia_prep` == `nvidia_test` and `nvidia_tests` == `nvidia_test_assertions`.

csibbitt replied:
Sure, I can rename the files as long as you don't mind that the tasks aren't literally `ansible.builtin.assert`ions. It was actually my original intention to convert them to asserts so that it's clear that there are no side effects, figuring we could make that a "rule" for this structure. Unfortunately the very first one I looked at (gpus_tests.yaml) is non-trivial to convert because it's checking the condition on each pass of the loop, so I'd need to change the code to loop a block, collect a list of outputs, and write a more complicated assertion to compare them to the config dictionary. Not a big deal, but altering the tests and potentially breaking them seemed out of scope for this PR. Ultimately, I'm not sure enforcing the asserts-only rule would be worth the effort in this case and potentially future cases where `failed_when` is clearer, but I could be convinced either way on that. Hopefully naming the file with `_assertions` is a clear enough indicator that it shouldn't have side effects.

csibbitt replied:
I dropped the _tests suffix on the latest commit, leaving the existing files just how they were, and have used _assertions be the new JUnit TEST-specific tasks. How's that?

lpiwowar replied:
Ok, this looks better from my point of view. But like I said, this was purely a subjective thing. Thank you for changing this!

lpiwowar commented on callback_plugins/custom_junit.py:
```diff
@@ -0,0 +1,34 @@
+
+        # Enforce our own defaults for when ENVVARs are inconvenient
+        self._output_dir = os.getenv('JUNIT_OUTPUT_DIR', '/var/lib/AnsibleTests/external_files/')
+        self._test_case_prefix = os.getenv('JUNIT_TEST_CASE_PREFIX', 'rhelai-validation : TEST')
+        self._fail_on_ignore = os.getenv('JUNIT_FAIL_ON_IGNORE', 'true')  # this is needed because we use "ignore_errors" on assertion tasks to run as many as possible
```
**question:**  I've noticed this in the Junit callback [1]. Does it mean that if an arbitrary task fails, it is going to be part of the junitXML report? Are we inheriting this behaviour? Do we want such a behavior? 

[1] https://github.com/ansible/ansible/blob/c02ad3516fced3c06486bd4b920e85f32e8e4f07/lib/ansible/plugins/callback/junit.py#L77C22-L77C149

csibbitt replied:
I think it does mean that, and we are inheriting it, and I don't know if it's harmful or not. The answer depends on whatever will parse this XML, and I have no real knowledge of that side yet. Personally I think it's a good thing, because if some test results are missing you can look in the file and see why. But I'm not an XML ingestion pipeline, so :shrug: 

I can at least test this and see what it looks like.

csibbitt replied:
Confirmed, they show up in the output, see my latest testing.

lpiwowar replied:
> But I'm not an XML ingestion pipeline, so 🤷

:smile: 

Sorry, for not providing the full context. The main reason why I mentioned this is that I very vaguely remember some jobs failing because a test case was not present in Polarion, and when `jump.py` tried to upload the results to Polarion the `jump.py` exited with an error code and the whole job failed. It was a long, long time ago. I remember seeing this in Jenkins. I haven't worked with it for quite some time now. 

I quickly checked the `jump.py` tool and it seems like there is a `--update-existing-test-cases` parameter which should ensure that `jump.py` will try to update test cases which are only present in Polarion.

I think we can mark this as resolved. What do you think? Because we have now control over the Junit callback it should be easily fixable in the future if it turns out it is a problem.

lpiwowar reviewed (commented):
Just a couple of thoughts that crossed my mind:).

csibbitt committed:
Restore existing names for _prep tasks, and use _assertions for TESTs

csibbitt committed:
Improved classname handling in custom_junit.py

csibbitt commented:
## Testing the classname + name parser

I used this test Playbook:

```
---
- name: Testing JUnit
  hosts: localhost
  gather_facts: false
  tasks:
    - name: TEST[sub1] Assert that should pass
      ansible.builtin.assert:
        that: true

    - name: TEST [sub2] Assert that should fail  # noqa: ignore-errors
      ansible.builtin.assert:
        that: false
      ignore_errors: true

    - name: TEST No class suffix
      ansible.builtin.assert:
        that: true

    - name: TEST     [$special[class]       W3i|2d%chars-----------------
      ansible.builtin.assert:
        that: true

    - name: Does not produce XML output
      ansible.builtin.assert:
        that: true

    - name: Does this produce XML output?
      ansible.builtin.assert:
        that: false
```

And it produced this XML
```
$ JUNIT_OUTPUT_DIR=. JUNIT_TEST_CASE_PREFIX=TEST ansible-playbook t.yaml

$ cat t-1741194716.7317407.xml
<?xml version="1.0" ?>
<testsuites disabled="0" errors="0" failures="2" tests="5" time="0.05054950714111328125">
	<testsuite disabled="0" errors="0" failures="2" name="t" skipped="0" tests="5" time="0.05054950714111328125">
		<testcase classname="rhoso_rhelai_validation.sub1" name="assert_that_should_pass" time="0.010919332504272461"/>
		<testcase classname="rhoso_rhelai_validation.sub2" name="assert_that_should_fail" time="0.00940561294555664">
			<failure message="Assertion failed" type="failure">{
    &quot;assertion&quot;: false,
    &quot;changed&quot;: false,
    &quot;evaluated_to&quot;: false,
    &quot;msg&quot;: &quot;Assertion failed&quot;
}</failure>
		</testcase>
		<testcase classname="rhoso_rhelai_validation" name="no_class_suffix" time="0.009761333465576172"/>
		<testcase classname="rhoso_rhelai_validation.special_class" name="w3i_2d_chars" time="0.009016990661621094"/>
		<testcase classname="rhoso_rhelai_validation" name="localhost_testing_junit_does_this_produce_xml_output" time="0.011446237564086914">
			<failure message="Assertion failed" type="failure">{
    &quot;assertion&quot;: false,
    &quot;changed&quot;: false,
    &quot;evaluated_to&quot;: false,
    &quot;msg&quot;: &quot;Assertion failed&quot;
}</failure>
		</testcase>
	</testsuite>
</testsuites>
```

You can see this also answers your question about a failed task with no prefix - yes, they end up in the XML. Like I said, this seems okay to me, but if it turns out to be a problem we can figure out how to exclude them.

Here is the output from an actual run of our playbooks using this callback
```
<?xml version="1.0" ?>
<testsuites disabled="0" errors="0" failures="0" tests="2" time="3.078845977783203125">
	<testsuite disabled="0" errors="0" failures="0" name="main" skipped="0" tests="2" time="3.078845977783203125">
		<testcase classname="rhoso_rhelai_validation.gpus" name="check_if_gpus_in_passthrough_mode_are_present_in_rhel_ai_vm_lspci" time="3.0535192489624023"/>
		<testcase classname="rhoso_rhelai_validation.nvidia" name="check_if_gpus_in_passthrough_mode_are_present_in_rhel_ai_vm_nvidia_smi" time="0.02532672882080078"/>
	</testsuite>
</testsuites>
```

csibbitt commented:
I'm done with this for now; please LMK if you'd like to see any more changes.

jpodivin commented on callback_plugins/custom_junit.py:
```diff
@@ -0,0 +1,47 @@
+
+            # Parse [<classname_suffix>] from begining of task name
+            class_suffix_pattern = r'^\s*\[(.*?)\]'
+            class_suffix_match = re.match(class_suffix_pattern, tc.name)
+            if (class_suffix_match):
```
You can combine two lines with a walrus operator https://docs.python.org/3/whatsnew/3.8.html#assignment-expressions

umago replied:
Yeah I really like that! That said, I think we can leave this as-is, it's just a nit

csibbitt replied:
Cool, thanks!

jpodivin reviewed (approved):
I have only exceedingly minor suggestion. Nothing that should block this, only something for future reference.

umago reviewed (approved):
Thanks Chris, this LGTM
